src/hdmf_zarr/zarr_utils.py

```diff
@@ -73,6 +74,16 @@ def __iter__(self):
     def __next__(self):
         return self._get_ref(super().__next__())
 
+    def append(self, arg):
+        # Building the parent first will build the new child.
+        # This correctly sets the parent of the child builder, which is needed to create the reference.
+        self.io.manager.build(arg.parent)
+        builder = self.io.manager.build(arg)
```

Comment:
On further thought, what if the parent of the new child has not been built? It may be safest to build the whole file then. Can you update the code to do that?

Would this have negative side effects? I can't think of any right now. but maybe if a user adds a container to the file, appends a reference elsewhere in the file, which would trigger a full build, then tries to remove the container or its parent from the file - the builder for the removed container will already have been built and attached. This is why usually all the building and reference-making happen during write() after which the user is not supposed to modify the in-memory objects anymore. 

I wonder if it would be better to defer the full file building to write() if that is possible, like we could store a proxy object here that gets processed during the io.write() call. Can you add a comment about this case in the code or create an issue ticket? We could try to address this now but for the sake of getting this out the door, I say we move forward and just do a full-file build here.